### DISCUSSION TITLE
Some Questions about event.listens_for and detecting changes

### DISCUSSION THREAD

**Discussion by radiantone** [sqlalchemy/sqlalchemy#8371] [Usage Questions] [answered]
Hi,
  Per the docs I try to determine if an object in the event callback has changed at all, and the suggested code to do this isn't returning true for me.

@event.listens_for(ProcessorModel, 'before_update')
def receive_before_update(mapper, connection, target):
....
    has_changes = object_session(target).is_modified(target, include_collections=False)
....

has_changes is False even when an attribute of target has changed from the stored version.

The other observation I had, is I can't seem to use a parent/base model class in the decorator. For example

@event.listens_for(BaseModel, 'before_update')

Doesn't trigger, even though this:

class ProcessorModel(HasLogs, BaseModel)

Thoughts? Tips? Thank you

**Comment by zzzeek** [comment, marked as answer]
has_changes is normally true if any local column attributes have changed. 

in more complex situations involving relationships, it's possible that the local attributes haven't changed at all, though would need to see specifics.  

making a plain vanilla test (which I recommend in order to isolate these behaviors) shows that for a simple case, it's True:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import object_session
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


@event.listens_for(A, "before_update")
def receive_before_update(mapper, connection, target):
    has_changes = object_session(target).is_modified(
        target, include_collections=False
    )
    print(f"has changes: {has_changes}")


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(data="d1"))
s.commit()

a1 = s.query(A).first()
a1.data = "d2"
s.flush()
```

```
SELECT a.id AS a_id, a.data AS a_data 
FROM a
 LIMIT ? OFFSET ?
2022-08-09 09:24:40,981 INFO sqlalchemy.engine.Engine [generated in 0.00008s] (1, 0)
has changes: True
2022-08-09 09:24:40,982 INFO sqlalchemy.engine.Engine UPDATE a SET data=? WHERE a.id = ?

```

> The other observation I had, is I can't seem to use a parent/base model class in the decorator. For example

make sure you pass [propagate=True](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=propagate#sqlalchemy.orm.MapperEvents.params.propagate) when listening on hierarchies

    @event.listens_for(BaseModel, 'before_update', propagate=True)

**Comment by radiantone** [comment]
Ok, thank you for the quick and detailed reply. I will give those suggestions a try and if it still doesn't work, I'll share more code. I am using postgres btw, so not sure if that would affect the is_modified call, but I can say that I am definitely changing a column in the object coming into the 'before_update' callback. This much I know. :)
